### DISCUSSION TITLE
Add Collection `ids` macro

### DISCUSSION THREAD

**Discussion by Anubarak** [craftcms/cms#12113] [Ideas] [answered]
Really simple:
similar to

https://github.com/craftcms/cms/blob/develop/src/base/ApplicationTrait.php#L1443
```PHP
Collection::macro('one', function() {
    /** @var Collection $this */
    return $this->first(...func_get_args());
});
```

it would be awesome to have a `ids` macro as well...
for example with the ArrayHelper to ensure compatibility with non Elements
```php
Collection::macro('ids', function() {
    /** @var Collection $this */
    return ArrayHelper::getColumn($this->all(), 'id');
});
```

The less `if $object instanceof Collection` checks wee need, the better

**Comment by brandonkelly** [comment, marked as answer]
Made this possible for Craft 4.3 via #12140.

Little more complicated than adding a new macro since we can’t assume every collection is going to contain items that have an `id` key/property.

Instead I added a new `craft\elements\ElementCollection` class, which `ElementQuery::collect()` and `Element::getEagerLoadedElements()` return instead of `Collection`.

Also went ahead and gave `ElementCollection` a `with()` method as well, so it’s now possible to eager-load nested elements after the fact, rather than having to specify everything before the element query is executed.

**Reply by brandonkelly** [reply]
Craft 4.3 was released earlier this week with those changes!

**Reply by gglnx** [reply]
@brandonkelly I think `one()` is missing too?

Slightly related (because I'm working with `phpstan`): the `Assets` field generates a wrong `valueType()` signature: In `\craft\elements\db\AssetQuery|\craft\elements\ElementCollection<craft\elements\Asset>` it's missing a leading backlash for the `ElementCollection` generic (resolving to: `craft\elements\db\AssetQuery|(craft\elements\ElementCollection&iterable<craft\behaviors\craft\elements\Asset>)` in phpstan). Should I open an issue for that?

**Reply by brandonkelly** [reply]
All collections have a `one()` macro that calls `first()` internally:

https://github.com/craftcms/cms/blob/bb51407401629d763f7936504769f10f336d803a/src/base/ApplicationTrait.php#L1459-L1463

That property type bug has been fixed for the next release (3f8a5d613dc2e7ab805041b187e76895ca12e12d). Please post any additional things you come across as new issues – thanks!

**Reply by gglnx** [reply]
@brandonkelly Maybe move it to `ElementCollection` (without the macro magic) or add an `@method` to `ElementCollection` so `phpstan` catches it? That would be great.

**Reply by brandonkelly** [reply]
Done (17fdc2afe932c9be30ef4c00df8b1d2055396da5).
